src/runtime/CollectionWrappers/IterableWrapper.cs

```diff
@@ -0,0 +1,42 @@
+using System;
+using System.Collections.Generic;
+using System.Collections;
+
+namespace Python.Runtime.CollectionWrappers
+{
+    internal class IterableWrapper<T> : IEnumerable<T>
+    {
+        protected readonly PyObject pyObject;
+
+        public IterableWrapper(PyObject pyObj)
+        {
+            if (pyObj == null)
+                throw new ArgumentNullException();
+            pyObject = pyObj;
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            PyObject iterObject = null;
+            using (Py.GIL())
+                iterObject = new PyObject(Runtime.PyObject_GetIter(pyObject.Handle));
+
+            while (true)
+            {
+                using (Py.GIL())
+                {
+                    var item = Runtime.PyIter_Next(iterObject.Handle);
+                    if (item == IntPtr.Zero)
+                    {
+                        iterObject.Dispose();
```

Comment:
You must check if there was an exception here.

https://docs.python.org/3.8/c-api/iter.html#c.PyIter_Next